hack/jenkins/test-history/gen_html.py

```diff
@@ -0,0 +1,118 @@
+#!/usr/bin/env python
+
+# Copyright 2016 The Kubernetes Authors All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Creates an HTML report for all jobs starting with a given prefix.
+
+Reads the JSON from tests.json, and prints the HTML to stdout.
+
+This code is pretty nasty, but gets the job done.
+
+It would be really spiffy if this used an HTML template system, but for now
+we're old-fashioned. We could also generate these with JS, directly from the
+JSON. That would allow custom filtering and stuff like that.
+"""
+
+from __future__ import print_function
+
+import json
+import string
+import sys
+
+def gen_tests(data, prefix):
+    """Creates the HTML for all test cases.
+
+    Returns the HTML along with the number of passing, unstable, failing, and
+    skipped tests.
+    """
+    html = '<ul class="test">\n'
+    total_okay = 0
+    total_unstable = 0
+    total_failed = 0
+    total_skipped = 0
+    for test in sorted(data, key=string.lower):
+        test_html = '<ul class="suite">\n'
+        has_test = False
+        has_failed = False
+        has_unstable = False
+        for suite in sorted(data[test]):
+            if not suite.startswith(prefix):
+                continue
+            has_test = True
+            num_failed = 0
+            num_builds = 0
+            total_time = 0
+            for build in data[test][suite]:
+                num_builds += 1
+                if build['failed']:
+                    num_failed += 1
+                total_time += build['time']
+            avg_time = total_time / num_builds
+            unit = 's'
+            if avg_time > 60:
+                avg_time /= 60
+                unit = 'm'
+            if num_failed > 0:
+                has_unstable = True
+            if num_failed == num_builds:
+                has_failed = True
+            status = 'okay' if num_failed == 0 else 'unstable' if num_failed < num_builds else 'failed'
+            test_html += '<li class="suite">'
+            test_html += '<span class="' + status + '">' + str(num_builds - num_failed) + '/' + str(num_builds) + '</span>'
+            test_html += '<span class="time">' + str(int(avg_time)) + unit + '</span>'
+            test_html += suite
+            test_html += '</li>\n'
+        status = 'failed' if has_failed else 'unstable' if has_unstable else 'okay' if has_test else 'skipped'
```

Comment:
I think this would be easier to read without the ternary operators and

if has_failed:
  status = failed
elif has_unstable:
  status = unstable
etc